### PR TITLE
[ADD] website_forum: moderation & anti spam

### PR DESCRIPTION
**Pull request by JulienLouette**
Better moderation / anti spam protection.
New "Moderation tools" box that hold the various moderation points,
displayed only if having at least "karma_moderation" karma.
New "Flag" link after "Delete" (available on karma_flag) when
displaying posts.
New "Validation queue", "Flagged queue" and "Offensive queue" to easily
moderate posts.

### CODE REVIEW

File: addons/website_forum/models/forum.py [hunk 1]
```diff
@@ -126,6 +126,8 @@ def _get_default_faq(self):
     karma_editor = fields.Integer(string='Editor Features: image and links',
                                   default=30, oldname='karma_editor_link_files')
     karma_user_bio = fields.Integer(string='Display detailed user biography', default=750)
+    karma_post = fields.Integer(string='Ask questions without validation', default=100)
```
Comment (tde-banana-odoo) [inline addons/website_forum/models/forum.py]: What is the different with the karma_ask and karma_answer fields? Is it anything preventing from using those fields ?

File: addons/website_forum/models/forum.py [hunk 2]
```diff
@@ -126,6 +126,8 @@ def _get_default_faq(self):
...
+    karma_moderate = fields.Integer(string='Moderate posts', default=1000)
```
Comment (tde-banana-odoo) [inline addons/website_forum/models/forum.py]: Small remark: we have a "karma_flag" at 500; maybe being able to moderate at 100 is a bit low :) .

File: addons/website_forum/models/forum.py [hunk 3]
```diff
@@ -193,7 +195,7 @@ def _get_plain_content(self):
 
     content_link = fields.Char('URL', help="URL of Link Articles")
     tag_ids = fields.Many2many('forum.tag', 'forum_tag_rel', 'forum_id', 'forum_tag_id', string='Tags')
-    state = fields.Selection([('active', 'Active'), ('close', 'Close'), ('offensive', 'Offensive')], string='Status', default='active')
+    state = fields.Selection([('active', 'Active'), ('pending', 'Waiting Validation'), ('close', 'Close'), ('offensive', 'Offensive'), ('flagged', 'Flagged')], string='Status', default='active')
```
Comment (tde-banana-odoo) [inline addons/website_forum/models/forum.py]: Is it necessary to differentiate offensive from flagged ? In my mind
someone posts with enough karma: active
someone posts without enough karma: pending
then someone validate: active
someone flags: offensive / flagged (choose)
someone reset the flag: active or close the posts: close (with a reason).

File: addons/website_forum/models/forum.py (hunk 1 above)
Comment (JulienLouette) [inline addons/website_forum/models/forum.py]: When you create a new user account on the forum, you have 0 as karma. When your email is validated, you have "karma_ask" karma, and you can ask a question on the forum, but until you don't have "karma_question", your question must be "accepted" by a moderator ("karma_moderator") to be displayed to all users ("trending", "newest", "search", ...).
Karma_answer is just the karma needed to answer a question. Those answers don't need to be "accepted" by a moderator (if you have "karma_answer", you can answer a question, and the answer is automatically displayed to all the users. We could change this if needed, but the idea is to don't have too many posts to moderate).

File: addons/website_forum/models/forum.py (hunk 2 above)
Comment (JulienLouette) [inline addons/website_forum/models/forum.py]: To be able to moderate, you need "karma_moderate" (1000). "karma_post" is just to be able to ask a question without waiting for a validation from a moderator.
But in my opinion, we could set a lower value for the "karma_flag", as a flag is just used to mark a post as "potentially offensive". A flagged post will then appear in a specific queue, sorted by number of flags, and managed by moderators. If the post is marked as offensive by a moderator, its new state is "offensive", and it is removed from the forum ("closed" for a question, hidden for an answer). If it is not offensive, its new state is "active" again, and the flags are removed.

File: addons/website_forum/models/forum.py (hunk 3 above)
Comment (JulienLouette) [inline addons/website_forum/models/forum.py]: What I do is just a little bit different. I distinguish "close" and "offensive". They have respectively different reasons to "close"/"hide" the post (for questions and answers).
Someone flags: "flagged" (many users can flag a same post, and a "flags_counter" is incremented).
A moderator can then mark the post as "offensive" (with an "offensive reason" (racism, insults, ...)), or "active".

**Comment by JKE-be** [comment]
:+1:

**Comment by tde-banana-odoo** [comment]
:postal_horn:

**Comment by JulienLouette** [comment]
Pad: https://pad.odoo.com/p/r.8eBngcxaS4V5JycS

File: addons/website_forum/models/forum.py [hunk 4]
```diff
@@ -226,6 +228,17 @@ def _compute_relevancy(self):
...
 
+    @api.one
+    @api.depends('flags_uid')
+    def _get_flags_count(self):
+        self.flags_count = len(self.flags_uid)
+
+    # moderation tools
+    validator_id = fields.Many2one('res.users', string='Reviewed by', select=True, readonly=True)
+    flag_moderator_id = fields.Many2one('res.users', string='Marked as offensive by', select=True, readonly=True)
+    flags_uid = fields.Many2many('res.users', relation='forum_post_flags_users_rel', string='Flagged by')
```
Comment (antonylesuisse) [inline addons/website_forum/models/forum.py]: flag_user_ids

Comment (antonylesuisse) [inline addons/website_forum/models/forum.py]: validator_id

flag_user_ids
flag_count
flag_moderator_id

**Pull request closed without merging**